src/server/editorServices.ts

```diff
@@ -541,174 +576,162 @@ namespace ts.server {
          */
-        private onConfigFileAddedForInferredProject(fileName: string) {
-            // TODO: check directory separators
-            if (getBaseFileName(fileName) !== "tsconfig.json") {
-                this.logger.info(`${fileName} is not tsconfig.json`);
-                return;
-            }
-
-            const { configFileErrors } = this.convertConfigFileContentToProjectOptions(fileName);
-            this.reportConfigFileDiagnostics(fileName, configFileErrors, fileName);
-
-            this.logger.info(`Detected newly added tsconfig file: ${fileName}`);
-            this.reloadProjects();
-        }
-
-        private getCanonicalFileName(fileName: string) {
-            const name = this.host.useCaseSensitiveFileNames ? fileName : fileName.toLowerCase();
-            return normalizePath(name);
+        private onConfigFileChangeForOpenScriptInfo(configFileName: NormalizedPath, eventKind: FileWatcherEventKind) {
+            // This callback is called only if we dont have config file project for this config file
+            const cononicalConfigPath = normalizedPathToPath(configFileName, this.currentDirectory, this.toCanonicalFileName);
+            const configFilePresenceInfo = this.mapOfConfigFilePresence.get(cononicalConfigPath);
+            configFilePresenceInfo.exists = (eventKind !== FileWatcherEventKind.Deleted);
+            this.logConfigFileWatchUpdate(configFileName, configFilePresenceInfo, ConfigFileWatcherStatus.ReloadingFiles);
+            // The tracking opens files would only contaion the inferred root so no need to check
+            this.delayReloadConfiguredProjectForFiles(configFilePresenceInfo.trackingOpenFilesMap, /*ignoreIfNotInferredProjectRoot*/ false);
         }
 
         private removeProject(project: Project) {
             this.logger.info(`remove project: ${project.getRootFiles().toString()}`);
 
             project.close();
+            // Remove the project from pending project updates
+            this.pendingProjectUpdates.delete(project.getProjectName());
 
             switch (project.projectKind) {
                 case ProjectKind.External:
                     removeItemFromSet(this.externalProjects, <ExternalProject>project);
                     this.projectToSizeMap.delete((project as ExternalProject).externalProjectName);
                     break;
                 case ProjectKind.Configured:
-                    removeItemFromSet(this.configuredProjects, <ConfiguredProject>project);
+                    this.configuredProjects.delete((<ConfiguredProject>project).canonicalConfigFilePath);
                     this.projectToSizeMap.delete((project as ConfiguredProject).canonicalConfigFilePath);
+                    this.setConfigFilePresenceByClosedConfigFile(<ConfiguredProject>project);
                     break;
                 case ProjectKind.Inferred:
                     removeItemFromSet(this.inferredProjects, <InferredProject>project);

```

Comment:
Does `projectToSizeMap` not count inferred projects? That should be documented at the declaration of `projectToSizeMap` if so.